src/icmp.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2019 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "icmp.h"
+
+#include "ip.h"
+#include "log.h"
+
+static
+uint16_t
+icmp6_cksum(struct ip6_hdr* nh, struct icmp6_hdr* th)
+{
+    uint32_t sum = 0;
+    uint16_t* w;
+    int i;
+
+    // IPv6 pseudo-header
+    //   - src and dst
+    w = (uint16_t*) &nh->ip6_src;
+    for (i = 0; i < 16; ++i)
+        sum += *w++;
+    sum += nh->ip6_plen;
+    sum += htons(nh->ip6_nxt);
+
+    // ICMP payload
+    w = (uint16_t*) th;
+    for (i = ntohs(nh->ip6_plen); i > 1; i -= 2)
+    {
+        sum += *w++;
+    }
+
+    if (i == 1)
+        sum += (uint16_t) *(uint8_t*)w;
+
+    sum += sum >> 16;
```

Comment:
I think this is wrong. Should be `while (sum>>16) sum = (sum>>16) + (sum&0xFFFF);` Consider what happens when sum = 0x2FFFF:

- When sum = 0x2FFFF, correct value for sum (before 1s complement) would be 0xFFFF + 0x2 = 0x10001, then 0x0001 + 0x1 = 0x0002. Checksum becomes ~0x0002 = 0xfffd.

With this code, sum becomes 0x30001. Checksum then becomes 0xfffe.

Reply:
Yes, thanks! I forgot to update this function based on your xbridge branch. Will fix.